test/cqlpy/test_filtering.py

```diff
@@ -152,6 +152,31 @@ def test_filtering_with_in_relation(cql, test_keyspace, cassandra_bug):
         res = cql.execute(f"select * from {table} where v in (5,7) ALLOW FILTERING")
         assert set(res) == set([(2,3,4,5), (4,5,6,7)])
 
+# Test that NOT IN restrictions are supported with filtering and return the
+# correct results.
+def test_filtering_with_not_in_relation(cql, test_keyspace):
+    schema = 'p1 int, p2 int, c int, v int, primary key ((p1, p2),c)'
+    with new_test_table(cql, test_keyspace, schema) as table:
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (1, 2, 3, 4)")
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (2, 3, 4, 5)")
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (3, 4, 5, 6)")
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (4, 5, 6, 7)")
+        cql.execute(f"INSERT INTO {table} (p1, p2, c, v) VALUES (5, 6, 7, NULL)")
+        res = cql.execute(f"select * from {table} where p1 NOT IN (2,4) ALLOW FILTERING")
+        assert set(res) == set([(1,2,3,4), (3,4,5,6), (5,6,7,None)])
+        res = cql.execute(f"select * from {table} where p2 NOT IN (2,4) ALLOW FILTERING")
+        assert set(res) == set([(2,3,4,5), (4,5,6,7), (5,6,7,None)])
+        res = cql.execute(f"select * from {table} where c NOT IN (3,5) ALLOW FILTERING")
+        assert set(res) == set([(2,3,4,5), (4,5,6,7), (5,6,7,None)])
+        res = cql.execute(f"select * from {table} where v NOT IN (5,7) ALLOW FILTERING")
+        assert set(res) == set([(1,2,3,4), (3,4,5,6)])
+        with pytest.raises(InvalidRequest, match='ALLOW FILTERING'):
+            cql.execute(f"select * from {table} where p1 NOT IN (2,4)")
+        # Since NULL is unknown, it could have matched p1, so NOT IN returns NULL too.
+        res = cql.execute(f"select * from {table} where v NOT IN (5,NULL) ALLOW FILTERING")
```

Comment:
I checked (the instructions in test/cqlpy/README on building Cassandra from source worked easily), and this line does not pass on today's Cassandra trunk.  It gives an error: "Invalid null value for column v". So please move this specific check to a separate test function and mark it scylla_only and explain in a comment the situation. Thanks.

Reply:
I forgot to git add the code that splits it into two.